From e5d69db1083481aef4ea64b504c294929c7422d9 Mon Sep 17 00:00:00 2001
From: "M. Vefa Bicakci" <vefa.bicakci@windriver.com>
Date: Fri, 14 Jan 2022 16:26:29 -0500
Subject: [PATCH] mlx5: pci_irq: Use irq_update_affinity_hint

This commit applies a patch that modifies the mlx5 driver so that it
uses the irq_update_affinity_hint function instead of the
irq_set_affinity_hint function. The former only sets the hint, whereas
the latter sets both the hint and the IRQ affinity.

The intent of the patch is to allow the user-specified IRQ affinity (via
the irqaffinity= command line argument) take effect for the IRQs set up
by the mlx5 device driver.

(Please see the description of the applied patch for more information.)

Signed-off-by: M. Vefa Bicakci <vefa.bicakci@windriver.com>
---
 ...pci_irq-Use-irq_update_affinity_hint.patch | 60 +++++++++++++++++++
 1 file changed, 60 insertions(+)
 create mode 100644 source/backports/1000-mlx5-pci_irq-Use-irq_update_affinity_hint.patch

diff --git a/source/backports/1000-mlx5-pci_irq-Use-irq_update_affinity_hint.patch b/source/backports/1000-mlx5-pci_irq-Use-irq_update_affinity_hint.patch
new file mode 100644
index 000000000000..f1122af264eb
--- /dev/null
+++ b/source/backports/1000-mlx5-pci_irq-Use-irq_update_affinity_hint.patch
@@ -0,0 +1,60 @@
+From 3884feaf05e9b1003ab83ab76fbfdf9a188c4a19 Mon Sep 17 00:00:00 2001
+From: "M. Vefa Bicakci" <vefa.bicakci@windriver.com>
+Date: Fri, 14 Jan 2022 16:26:29 -0500
+Subject: [PATCH] mlx5: pci_irq: Use irq_update_affinity_hint
+
+The StarlingX kernel was patched to deprecate irq_set_affinity_hint
+by cherry-picking the patches at:
+  https://lore.kernel.org/netdev/20210903152430.244937-1-nitesh@redhat.com/t/#u
+
+These patches have been mainlined as of this writing, with the following
+merge commit by Linus Torvalds:
+  https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=147cc5838c0f5c76e908b816e924ca378e0d4735
+
+This commit modifies the mlx5 driver so that it uses the
+irq_update_affinity_hint function instead of the irq_set_affinity_hint
+function. The former only sets the hint, whereas the latter sets both
+the hint and the IRQ affinity.
+
+Please note that this is a divergence from the aforementioned patch
+series, which make mlx5 use irq_set_affinity_and_hint, which currently
+behaves in the same way as irq_set_affinity_hint. The intent with
+diverging from mainline is to allow the user-specified IRQ affinity (via
+the irqaffinity= command line argument) take effect for the IRQs
+set up by the mlx5 device driver.
+
+The mlx5 commit in mainline is accessible at:
+  https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=7451e9ea8e2055af39afe7ff39a5f68d8ec6b98d
+
+Signed-off-by: M. Vefa Bicakci <vefa.bicakci@windriver.com>
+---
+ drivers/net/ethernet/mellanox/mlx5/core/pci_irq.c | 6 +++---
+ 1 file changed, 3 insertions(+), 3 deletions(-)
+
+diff --git a/drivers/net/ethernet/mellanox/mlx5/core/pci_irq.c b/drivers/net/ethernet/mellanox/mlx5/core/pci_irq.c
+index 09d0ce8061f3..db7472d10fb6 100644
+--- a/drivers/net/ethernet/mellanox/mlx5/core/pci_irq.c
++++ b/drivers/net/ethernet/mellanox/mlx5/core/pci_irq.c
+@@ -237,8 +237,8 @@ static int set_comp_irq_affinity_hint(struct mlx5_core_dev *mdev, int i)
+ 	cpumask_set_cpu(cpumask_local_spread(i, mdev->priv.numa_node),
+ 			irq->mask);
+ 	if (IS_ENABLED(CONFIG_SMP) &&
+-	    irq_set_affinity_hint(irqn, irq->mask))
+-		mlx5_core_warn(mdev, "irq_set_affinity_hint failed, irq 0x%.4x",
++	    irq_update_affinity_hint(irqn, irq->mask))
++		mlx5_core_warn(mdev, "irq_update_affinity_hint failed, irq 0x%.4x",
+ 			       irqn);
+ 
+ 	return 0;
+@@ -261,7 +261,7 @@ static void clear_comp_irq_affinity_hint(struct mlx5_core_dev *mdev, int i)
+ 		msix = priv->msix_arr;
+ 		irqn                 = msix[vecidx].vector;
+ #endif
+-	irq_set_affinity_hint(irqn, NULL);
++	irq_update_affinity_hint(irqn, NULL);
+ 	free_cpumask_var(irq->mask);
+ }
+ 
+-- 
+2.29.2
+
-- 
2.29.2

